neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,443 @@
+    /// sent. During probing, this may be smaller than the actual path MTU.
+    #[must_use]
+    pub fn plpmtu(&self) -> usize {
+        self.mtu - self.header_size
+    }
+
+    /// Returns true if a PMTUD probe should be sent.
+    #[must_use]
+    pub fn needs_probe(&self) -> bool {
+        self.probe_state == Probe::Needed
+    }
+
+    /// Returns the size of the current PMTUD probe.
+    fn probe_size(&self) -> usize {
+        self.search_table[self.probe_index] - self.header_size
+    }
+
+    /// Sends a PMTUD probe.
+    pub fn send_probe(
+        &mut self,
+        builder: &mut PacketBuilder,
+        stats: &mut Stats,
+        aead_expansion: usize,
+    ) {
+        builder.set_limit(self.probe_size() - aead_expansion);
+        // The packet may include ACK-eliciting data already, but rather than check for that, it
+        // seems OK to burn one byte here to simply include a PING.
+        builder.encode_varint(FRAME_TYPE_PING);
+        builder.enable_padding(true);
+        stats.frame_tx.ping += 1;
+        stats.frame_tx.all += 1;
+        stats.pmtud_tx += 1;
+        self.probe_count += 1;
+        self.probe_state = Probe::Sent;
+        qdebug!(
+            "Sending PMTUD probe of size {}, count {}",
+            self.search_table[self.probe_index],
+            self.probe_count
+        );
+    }
+
+    /// Returns true if the packet is a PMTUD probe.
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent && p.len() == self.probe_size()
+    }
+
+    /// Count the PMTUD probes included in `pkts`.
+    fn count_pmtud_probes(&self, pkts: &[SentPacket]) -> usize {
+        pkts.iter().filter(|p| self.is_pmtud_probe(p)).count()
+    }
+
+    /// Checks whether a PMTUD probe has been acknowledged, and if so, updates the PMTUD state.
+    /// May also initiate a new probe process for a larger MTU.
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        let acked = self.count_pmtud_probes(acked_pkts);
+        if acked == 0 {
+            return;
+        }
+        // A probe was ACKed, confirm the new MTU and try to probe upwards further.
+        stats.pmtud_ack += acked;
+        self.mtu = self.search_table[self.probe_index];
+        qdebug!("PMTUD probe of size {} succeeded", self.mtu);
+        self.start_pmtud();
+    }
+
+    /// Stops the PMTUD process, setting the MTU to the largest successful probe size.
+    fn stop_pmtud(&mut self, idx: usize, now: Instant) {
+        self.probe_state = Probe::NotNeeded; // We don't need to send any more probes
+        self.probe_index = idx; // Index of the last successful probe
+        self.mtu = self.search_table[idx]; // Leading to this MTU
+        self.probe_count = 0; // Reset the count
+        self.loss_counts.fill(0); // Reset the loss counts
+        self.raise_timer = Some(now + PMTU_RAISE_TIMER);
```

Comment:
I think the argument for a fixed time is that PMTU changes would be due to routing changes, which are not correlated with the RTT. Whether 10 min is a good number is a valid question though.

Reply:
The reason I suggest this is that we don't have any fixed time values in the stack thus far.

This is the second number that is divorced from RTT.  The other is the idle timeout.  For that, we have a connection parameter.  Perhaps this needs one too.  (I'm thinking about our crazy RTT tests, where this will definitely affect things in surprising ways.)

Reply:
If we wanted to make this per-RTT, we could roughly say that 10min = 12000 "standard" RTTs @ 50ms, and just work with some sort of RTT count like that?

Reply:
If we want to make this change, let's do it in a follow-on PR.